Avoid using UTF-8 characters when printing in install

--- systemd.orig/src/shared/install.c
+++ systemd/src/shared/install.c
@@ -347,8 +347,8 @@ static void install_change_dump_success(
         switch (change->type) {
 
         case INSTALL_CHANGE_SYMLINK:
-                return log_info("Created symlink '%s' %s '%s'.",
-                                change->path, special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), change->source);
+                return log_info("Created symlink '%s' -> '%s'.",
+                                change->path, change->source);
 
         case INSTALL_CHANGE_UNLINK:
                 return log_info("Removed '%s'.", change->path);
@@ -603,8 +603,8 @@ static int create_symlink(
         }
 
         if (chroot_unit_symlinks_equivalent(lp, new_path, dest, old_path)) {
-                log_debug("Symlink %s %s %s already exists",
-                          new_path, special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), dest);
+                log_debug("Symlink %s -> %s already exists",
+                          new_path, dest);
                 return 1;
         }
 
