--- a/src/shared/install.c
+++ b/src/shared/install.c
@@ -320,9 +320,8 @@
                 switch(changes[i].type) {
                 case UNIT_FILE_SYMLINK:
                         if (!quiet)
-                                log_info("Created symlink %s %s %s.",
+                                log_info("Created symlink %s -> %s.",
                                          changes[i].path,
-                                         special_glyph(SPECIAL_GLYPH_ARROW),
                                          changes[i].source);
                         break;
                 case UNIT_FILE_UNLINK:
