Index: os-systemd/src/shared/install.c
===================================================================
--- os-systemd.orig/src/shared/install.c
+++ os-systemd/src/shared/install.c
@@ -351,9 +351,8 @@ void install_changes_dump(int r, const c
                 switch (changes[i].type) {
                 case INSTALL_CHANGE_SYMLINK:
                         if (!quiet)
-                                log_info("Created symlink %s %s %s.",
+                                log_info("Created symlink %s -> %s.",
                                          changes[i].path,
-                                         special_glyph(SPECIAL_GLYPH_ARROW_RIGHT),
                                          changes[i].source);
                         break;
                 case INSTALL_CHANGE_UNLINK:
@@ -548,8 +547,8 @@ static int create_symlink(
         }
 
         if (chroot_unit_symlinks_equivalent(lp, new_path, dest, old_path)) {
-                log_debug("Symlink %s %s %s already exists",
-                          new_path, special_glyph(SPECIAL_GLYPH_ARROW_RIGHT), dest);
+                log_debug("Symlink %s -> %s already exists",
+                          new_path, dest);
                 return 1;
         }
 
