Index: os-systemd/src/core/path.c
===================================================================
--- os-systemd.orig/src/core/path.c
+++ os-systemd/src/core/path.c
@@ -197,10 +197,15 @@ int path_spec_fd_event(PathSpec *s, uint
                 return log_error_errno(errno, "Failed to read inotify event: %m");
         }
 
-        if (IN_SET(s->type, PATH_CHANGED, PATH_MODIFIED))
-                FOREACH_INOTIFY_EVENT_WARN(e, buffer, l)
+        if (IN_SET(s->type, PATH_CHANGED, PATH_MODIFIED)) {
+                FOREACH_INOTIFY_EVENT_WARN(e, buffer, l) {
+                        if (e->wd == -1) {
+                                log_error("%s: Detected event queue overflow!", __func__);
+                        }
                         if (s->primary_wd == e->wd)
                                 return 1;
+                }
+        }
 
         return 0;
 }
