diff --git a/src/core/path.c b/src/core/path.c
index ca3a91d..84a18b3 100644
--- a/src/core/path.c
+++ b/src/core/path.c
@@ -166,6 +166,9 @@ int path_spec_fd_event(PathSpec *s, uint32_t revents) {
         }
 
         FOREACH_INOTIFY_EVENT(e, buffer, l) {
+                if (e->wd == -1) {
+                        log_error("%s: Detected event queue overflow!", __func__);
+                }
                 if (IN_SET(s->type, PATH_CHANGED, PATH_MODIFIED) &&
                     s->primary_wd == e->wd)
                         r = 1;
