--- a/src/core/path.c
+++ b/src/core/path.c
@@ -175,6 +175,9 @@
         }
 
         FOREACH_INOTIFY_EVENT(e, buffer, l) {
+                if (e->wd == -1) {
+                        log_error("%s: Detected event queue overflow!", __func__);
+                }
                 if (IN_SET(s->type, PATH_CHANGED, PATH_MODIFIED) &&
                     s->primary_wd == e->wd)
                         r = 1;
