Index: os-systemd/src/core/path.c
===================================================================
--- os-systemd.orig/src/core/path.c
+++ os-systemd/src/core/path.c
@@ -26,6 +26,13 @@
 #include "unit-name.h"
 #include "unit.h"
 
+typedef enum PathWaitingSource {
+       PATH_WAITING_SOURCE_START,
+       PATH_WAITING_SOURCE_COLDPLUG,
+       PATH_WAITING_SOURCE_DISPATCH_IO,
+       PATH_WAITING_SOURCE_TRIGGER_NOTIFY,
+} PathWaitingSource;
+
 static const UnitActiveState state_translation_table[_PATH_STATE_MAX] = {
         [PATH_DEAD]    = UNIT_INACTIVE,
         [PATH_WAITING] = UNIT_ACTIVE,
@@ -198,7 +205,7 @@ int path_spec_fd_event(PathSpec *s, uint
         return 0;
 }
 
-static bool path_spec_check_good(PathSpec *s, bool initial, bool from_trigger_notify, char **ret_trigger_path) {
+static bool path_spec_check_good(PathSpec *s, PathWaitingSource source, char **ret_trigger_path) {
         _cleanup_free_ char *trigger = NULL;
         bool b, good = false;
 
@@ -226,7 +233,11 @@ static bool path_spec_check_good(PathSpe
         case PATH_CHANGED:
         case PATH_MODIFIED:
                 b = access(s->path, F_OK) >= 0;
-                good = !initial && !from_trigger_notify && b != s->previous_exists;
+                good = (source != PATH_WAITING_SOURCE_START) &&
+                       (source != PATH_WAITING_SOURCE_TRIGGER_NOTIFY) &&
+                       (source != PATH_WAITING_SOURCE_COLDPLUG ||
+                                s->previous_exists_deserialized) &&
+                        b != s->previous_exists;
                 s->previous_exists = b;
                 break;
 
@@ -480,7 +491,7 @@ static void path_set_state(Path *p, Path
         unit_notify(UNIT(p), state_translation_table[old_state], state_translation_table[state], 0);
 }
 
-static void path_enter_waiting(Path *p, bool initial, bool from_trigger_notify);
+static void path_enter_waiting(Path *p, PathWaitingSource source);
 
 static int path_coldplug(Unit *u) {
         Path *p = PATH(u);
@@ -491,7 +502,7 @@ static int path_coldplug(Unit *u) {
         if (p->deserialized_state != p->state) {
 
                 if (IN_SET(p->deserialized_state, PATH_WAITING, PATH_RUNNING))
-                        path_enter_waiting(p, true, false);
+                        path_enter_waiting(p, PATH_WAITING_SOURCE_COLDPLUG);
                 else
                         path_set_state(p, p->deserialized_state);
         }
@@ -561,18 +572,18 @@ fail:
         path_enter_dead(p, PATH_FAILURE_RESOURCES);
 }
 
-static bool path_check_good(Path *p, bool initial, bool from_trigger_notify, char **ret_trigger_path) {
+static bool path_check_good(Path *p, PathWaitingSource source, char **ret_trigger_path) {
         assert(p);
         assert(ret_trigger_path);
 
         LIST_FOREACH(spec, s, p->specs)
-                if (path_spec_check_good(s, initial, from_trigger_notify, ret_trigger_path))
+                if (path_spec_check_good(s, source, ret_trigger_path))
                         return true;
 
         return false;
 }
 
-static void path_enter_waiting(Path *p, bool initial, bool from_trigger_notify) {
+static void path_enter_waiting(Path *p, PathWaitingSource source) {
         _cleanup_free_ char *trigger_path = NULL;
         Unit *trigger;
         int r;
@@ -588,7 +599,7 @@ static void path_enter_waiting(Path *p,
                 return;
         }
 
-        if (path_check_good(p, initial, from_trigger_notify, &trigger_path)) {
+        if (path_check_good(p, source, &trigger_path)) {
                 log_unit_debug(UNIT(p), "Got triggered.");
                 path_enter_running(p, trigger_path);
                 return;
@@ -598,11 +609,15 @@ static void path_enter_waiting(Path *p,
         if (r < 0)
                 goto fail;
 
-        /* Hmm, so now we have created inotify watches, but the file
-         * might have appeared/been removed by now, so we must
-         * recheck */
+         /* Hmm, so now we have created inotify watches, but the file
+          * might have appeared/been removed by now, so we must
+          * recheck. Ensure, however, that we do not treat the recheck
+          * as coming from either an initial or coldplug source. */
+
+        if (source != PATH_WAITING_SOURCE_TRIGGER_NOTIFY)
+                source = PATH_WAITING_SOURCE_DISPATCH_IO;
 
-        if (path_check_good(p, false, from_trigger_notify, &trigger_path)) {
+        if (path_check_good(p, source, &trigger_path)) {
                 log_unit_debug(UNIT(p), "Got triggered.");
                 path_enter_running(p, trigger_path);
                 return;
@@ -644,7 +659,7 @@ static int path_start(Unit *u) {
         path_mkdir(p);
 
         p->result = PATH_SUCCESS;
-        path_enter_waiting(p, true, false);
+        path_enter_waiting(p, PATH_WAITING_SOURCE_START);
 
         return 1;
 }
@@ -741,6 +756,7 @@ static int path_deserialize_item(Unit *u
                                     path_equal(s->path, unescaped)) {
 
                                         s->previous_exists = previous_exists;
+                                        s->previous_exists_deserialized = true;
                                         break;
                                 }
                 }
@@ -795,7 +811,7 @@ static int path_dispatch_io(sd_event_sou
         if (changed)
                 path_enter_running(p, found->path);
         else
-                path_enter_waiting(p, false, false);
+                path_enter_waiting(p, PATH_WAITING_SOURCE_DISPATCH_IO);
 
         return 0;
 
@@ -861,7 +877,7 @@ static void path_trigger_notify_impl(Uni
                 return;
 
         if (on_defer) {
-                path_enter_waiting(p, /* initial = */ false, /* from_trigger_notify = */ true);
+                path_enter_waiting(p, PATH_WAITING_SOURCE_TRIGGER_NOTIFY);
                 return;
         }
 
Index: os-systemd/src/core/path.h
===================================================================
--- os-systemd.orig/src/core/path.h
+++ os-systemd/src/core/path.h
@@ -31,6 +31,7 @@ typedef struct PathSpec {
         int primary_wd;
 
         bool previous_exists;
+        bool previous_exists_deserialized;
 } PathSpec;
 
 int path_spec_watch(PathSpec *s, sd_event_io_handler_t handler);
