--- a/src/random-seed/random-seed.c
+++ b/src/random-seed/random-seed.c
@@ -2,7 +2,9 @@
 
 #include <errno.h>
 #include <fcntl.h>
+#include <linux/random.h>
 #include <string.h>
+#include <sys/ioctl.h>
 #include <sys/stat.h>
 #include <unistd.h>
 
@@ -20,10 +22,42 @@
 #define POOL_SIZE_MIN 512
 #define POOL_SIZE_MAX (10*1024*1024)
 
+static ssize_t read_entropy(int fd, struct rand_pool_info *info) {
+        ssize_t k;
+        int r;
+
+        assert(fd >= 0);
+        assert(info);
+
+        r = ioctl(fd, RNDGETENTCNT, &info->entropy_count);
+        if (r < 0)
+                return -errno;
+
+        k = loop_read(fd, &info->buf, info->buf_size, false);
+        if (k < 0)
+                return k;
+        info->buf_size = k;
+
+        return k + sizeof(struct rand_pool_info);
+}
+
+static int write_entropy(int fd, struct rand_pool_info *info) {
+        int r;
+
+        assert(fd >= 0);
+        assert(info);
+
+        r = ioctl(fd, RNDADDENTROPY, info);
+        if (r < 0)
+                return -errno;
+
+        return 0;
+}
+
 static int run(int argc, char *argv[]) {
         _cleanup_close_ int seed_fd = -1, random_fd = -1;
         bool read_seed_file, write_seed_file;
-        _cleanup_free_ void* buf = NULL;
+        _cleanup_free_ struct rand_pool_info* buf = NULL;
         size_t buf_size = 0;
         struct stat st;
         ssize_t k;
@@ -114,10 +148,13 @@
         if ((uint64_t) st.st_size > buf_size)
                 buf_size = MIN(st.st_size, POOL_SIZE_MAX);
 
-        buf = malloc(buf_size);
+        buf = malloc(buf_size + sizeof(struct rand_pool_info));
         if (!buf)
                 return log_oom();
 
+        buf->buf_size = buf_size;
+        buf_size += sizeof(struct rand_pool_info);
+
         if (read_seed_file) {
                 sd_id128_t mid;
                 int z;
@@ -131,9 +168,14 @@
                 } else {
                         (void) lseek(seed_fd, 0, SEEK_SET);
 
-                        r = loop_write(random_fd, buf, (size_t) k, false);
-                        if (r < 0)
-                                log_error_errno(r, "Failed to write seed to /dev/urandom: %m");
+                        if (k < sizeof(struct rand_pool_info) || k < buf->buf_size) {
+                                log_error("Seed file " RANDOM_SEED "corrupt or pool size shrank.");
+                        } else {
+                                r = write_entropy(random_fd, buf);
+                                if (r < 0)
+                                        log_error_errno(r, "Failed to write seed to /dev/urandom: %m");
+                        }
+
                 }
 
                 /* Let's also write the machine ID into the random seed. Why? As an extra protection against "golden
@@ -159,7 +201,7 @@
                 (void) fchmod(seed_fd, 0600);
                 (void) fchown(seed_fd, 0, 0);
 
-                k = loop_read(random_fd, buf, buf_size, false);
+                k = read_entropy(random_fd, buf);
                 if (k < 0)
                         return log_error_errno(k, "Failed to read new seed from /dev/urandom: %m");
                 if (k == 0)
